### PR TITLE
Add notification in renderer if lens protocol handler fails to find any routes

### LINKED ISSUE #2747: Exception Handling for Protocol Handler
**What would you like to be added**:
When a Lens URL is not correct have a way to hand the exceptions

**Why is this needed**:
Currently Lens comes to the foreground but nothing happens. The user could be confused

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #2747 

Some screenshots:
Screen Shot 2021-05-17 at 12 15 57 PM
Screen Shot 2021-05-17 at 12 16 14 PM
Screen Shot 2021-05-17 at 12 28 04 PM

### CODE REVIEW

File: src/renderer/protocol-handler/app-handlers.tsx [hunk 1]
```diff
@@ -30,14 +31,25 @@ import { entitySettingsURL } from "../components/+entity-settings";
...
     .getInstance()
     .addInternalHandler("/preferences", ({ search: { highlight }}) => {
       navigate(preferencesURL({ fragment: highlight }));
     })
-    .addInternalHandler("/", () => {
+    .addInternalHandler("/", ({ tail }) => {
+      if (tail) {
+        Notifications.info((
+          <p>
+            Missing handler for <code>lens://app/{tail}</code>. Are you on the latest version?
```
Comment (jim-docker) [inline src/renderer/protocol-handler/app-handlers.tsx]: Handlers might be removed in a "latest" version too, right? "correct" might be better than "latest"

**Review by jim-docker — COMMENTED** [review]
There's no notification if I specify the wrong cluster id, like
```
open lens://app/cluster/minikube
```
but it is reported in the logs:
```
[0] info: [PROTOCOL ROUTER]: routing lens://app/cluster/minikube
[0] info: [PROTOCOL ROUTER]: No handler found {"url":"lens://app/cluster/minikube"}
```

**Comment by Nokel81** [comment]
Those logs I assume are in `main`, so that is expected.

As for the renderer side, sure good idea.

**Comment by jim-docker** [comment]
> As for the renderer side, sure good idea.

I guess it's not actually a failure to find a route, but from the user's perspective it's the same thing: no navigation, and no failure notification.

**Comment by Nokel81** [comment]
I think you are confused by the log line. What it is reporting is that it cannot find a handler for the route `lens://app/cluster/minikube` on the `main` side.

On the `renderer` side it does find a handler and you should see a log line there saying: 

```
[APP-HANDLER]: cluster with given ID does not exist { clusterId: "minikube" }
```

**Comment by jim-docker** [comment]
Yes, see above

File: src/renderer/protocol-handler/app-handlers.tsx (hunk 1 above)
Comment (jim-docker) [inline src/renderer/protocol-handler/app-handlers.tsx]: This looks like a log message. The user probably doesn't know/care about handlers, just that the action couldn't be completed.

File: src/renderer/protocol-handler/router.tsx [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      source: ipcRenderer,
+      verifier: verifyIpcArgs,
+      listener: async (event, rawUrl, mainAttemptResult) => {
+        const rendererAttempt = await this._routeToExtension(new Url(rawUrl, true));
+
+        switch (foldAttemptResults(mainAttemptResult, rendererAttempt)) {
+          case RouteAttempt.MISSING:
+            Notifications.info((
+              <p>
+                Missing handler for <code>{rawUrl}</code>. Are you on the latest version of the extensions?
```
Comment (jim-docker) [inline src/renderer/protocol-handler/router.tsx]: Do you mean
```suggestion
                Missing handler for <code>{rawUrl}</code>. Are you on the latest version of the extension?
```

File: src/main/catalog-sources/kubeconfig-sync.ts [hunk 3]
```diff
@@ -68,28 +67,26 @@ export class KubeconfigSyncManager extends Singleton {
...
+        for (const filePath of this.sources.keys()) {
+          this.stopOldSync(filePath);
+        }
 
-    catalogEntityRegistry.removeSource(KubeconfigSyncManager.syncName);
-    this.syncing = false;
+        catalogEntityRegistry.removeSource(KubeconfigSyncManager.syncName);
+        this.syncing = false;
+      }
+    );
```
Comment (jim-docker) [inline src/main/catalog-sources/kubeconfig-sync.ts]: Are these changes related to protocol handler error notification?

File: src/main/cluster-manager.ts [hunk 4]
```diff
@@ -35,29 +35,36 @@ export class ClusterManager extends Singleton {
...
+      }),
+      () => {
+        for (const cluster of ClusterStore.getInstance().clusters.values()) {
+          cluster.disconnect();
+        }
       }
-    }, {
-      delay: 250
-    });
+    );
```
Comment (jim-docker) [inline src/main/cluster-manager.ts]: And this? Maybe you should've done an "Add `disposers()` to Singleton" PR first. Can you add this to the PR title/description?

File: src/main/exit-app.ts [hunk 5]
```diff
@@ -26,19 +26,11 @@ import { ClusterManager } from "./cluster-manager";
...
-  clusterManager?.stop();
+
+  WindowManager.resetInstance();
+  ClusterManager.resetInstance();
   logger.info("SERVICE:QUIT");
-  setTimeout(() => {
-    app.exit();
-  }, 1000);
+
+  setTimeout(() => app.exit(), 1000);
```
Comment (jim-docker) [inline src/main/exit-app.ts]: And this?

File: src/renderer/protocol-handler/app-handlers.tsx (hunk 1 above)
Comment (Nokel81) [reply src/renderer/protocol-handler/app-handlers.tsx]: How about "Unknown action"?

File: src/main/cluster-manager.ts (hunk 4 above)
Comment (Nokel81) [reply src/main/cluster-manager.ts]: Probably a good idea, will do that instead.

File: src/renderer/protocol-handler/app-handlers.tsx (hunk 1 above)
Comment (Nokel81) [reply src/renderer/protocol-handler/app-handlers.tsx]: Changed to "Unknown action"

Comment (jim-docker) [reply src/renderer/protocol-handler/app-handlers.tsx]: Should you change it here too?

**Review by jim-docker — COMMENTED** [review]
Were you going to add a notification when the cluster id can't be found?
i.e. for 
```
open lens://app/cluster/minkube
```

File: src/renderer/components/notifications/notifications.tsx [hunk 6]
```diff
@@ -37,20 +37,27 @@ export class Notifications extends React.Component {
   static ok(message: NotificationMessage) {
     notificationsStore.add({
       message,
-      timeout: 2500,
+      timeout: 2_500,
```
Comment (Nachasic) [inline src/renderer/components/notifications/notifications.tsx:40]: There is an ESLint plugin that adds a `numeric-separators-style` rule that could be useful, I think
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/numeric-separators-style.md

**Review by Nachasic — DISMISSED** [review]
LGMT, a couple of minor comments

File: src/renderer/components/notifications/notifications.tsx (hunk 6 above)
Comment (Nokel81) [reply src/renderer/components/notifications/notifications.tsx:40]: I think that should be a separate PR.

**Review by jim-docker — APPROVED** [review]

File: src/main/index.ts [hunk 7]
```diff
@@ -93,8 +93,11 @@ if (!app.requestSingleInstanceLock()) {
 
   for (const arg of process.argv) {
     if (arg.toLowerCase().startsWith("lens://")) {
-      lprm.route(arg)
-        .catch(error => logger.error(`${LensProtocolRouterMain.LoggingPrefix}: an error occured`, { error, rawUrl: arg }));
+      try {
```
Comment (jakolehm) [inline src/main/index.ts:96]: Should we refactor this logic to a function? I'm seeing exactly the same code on lines 108-116.

**Review by jakolehm — APPROVED** [review]

File: src/main/index.ts (hunk 7 above)
Comment (Nokel81) [reply src/main/index.ts:96]: Sure, I'll do that in a follow up PR.
